File: tools/ipv4_deduplication_aggregation.py
import argparse
from functools import reduce
from ipaddress import IPv4Network, ip_network
from sys import exit
from typing import (Any, Callable, Generic, List, NoReturn, Optional, Tuple,
                    TypeVar, Union)


Value = TypeVar("Value", List[str], List[IPv4Network])
Error = TypeVar("Error", str, Exception, contravariant=True)


class Result(Generic[Error, Value]):
    @property
    def value(self):
        raise NotImplementedError()

    @property
    def error(self):
        raise NotImplementedError()

    def fmap(self, func: Callable[[Value], Any]) -> "Result":
        raise NotImplementedError()

    def bind(self, func: Callable[[Value], Any]) -> "Result":
        raise NotImplementedError()

    def is_failure(self) -> bool:
        return isinstance(self, Failure)

    def is_success(self) -> bool:
        return isinstance(self, Success)


class Success(Result):
    @property
    def value(self) -> Union[List[str], List[IPv4Network]]:
        return self._value

    def __init__(self, value: Value):
        self._value: Union[List[str], List[IPv4Network]] = value

    def fmap(self, func: Callable[[Union[List[str], List[IPv4Network]]], Any]) -> "Success":
        assert callable(func), "fmap argument has to be callable"

        return Success(func(self._value))

    def bind(self, func: Callable[[Union[List[str], List[IPv4Network]]], Result[Error, Value]]) -> Result[Error, Value]:
        return func(self._value)


class Failure(Result):
    @property
    def error(self) -> Union[str, Exception]:
        return self._error

    def __init__(self, error: Error):
        self._error: Union[str, Exception] = error

    def fmap(self, _: Any) -> "Failure":
        return self

    def bind(self, _: Any) -> "Failure":
        return self


def get_nets_from_input(string: Optional[str], filepath: Optional[str]) -> Result[Error, List[str]]:
    source: List[str]
    error: str

    if string is None and filepath is None:
        return Failure("No input defined. Choose string or file input.")
    if string is not None and filepath is not None:
        return Failure("Both input options can not be used at the same time.")
    if string is not None:
        source = string.strip().split(" ")
        error = f"string: `{string}`"
    else:
        try:
            with open(filepath, "r") as file:  # type: ignore
                source = [l.strip() for l in file.readlines()]
                error = f"file: `{filepath}`"

        except (FileNotFoundError, PermissionError) as exc:
            return Failure(exc)

    lines: List[str] = [l for l in set(source) if l and l.isprintable()]

    return Success(lines) if lines else Failure(f"It seems given {error} does not contain a proper data.")


def get_net_from_str(nets: List[str]) -> Result[Error, List[IPv4Network]]:
    try:
        return Success(sorted(set(ip_network(n) for n in nets)))

    except ValueError as exc:
        return Failure(exc)


def aggregate_networks(nets: List[IPv4Network]) -> List[IPv4Network]:

    def large_absorb_small(aggregated_supernet: Tuple[List[IPv4Network], IPv4Network], net: IPv4Network) -> Tuple[List[IPv4Network], IPv4Network]:
        aggregated, supernet = aggregated_supernet

        if not net.subnet_of(supernet):  # type: ignore
            aggregated.append(net)
            supernet = net

        return aggregated, supernet

    def small_merge_large(aggregated_net1: Tuple[List[IPv4Network], IPv4Network], net2: IPv4Network) -> Tuple[List[IPv4Network], IPv4Network]:
        aggregated, net1 = aggregated_net1
        net1_supernet = net1.supernet()

        if net1.prefixlen == net2.prefixlen and net1_supernet == net2.supernet():
            aggregated.append(net1_supernet)
        else:
            if not aggregated:
                aggregated.append(net1)
            elif aggregated[-1] != net1 and aggregated[-1] != net1_supernet:
                aggregated.append(net1)
            aggregated.append(net2)

        return aggregated, net2

    if len(nets) == 1:
        return nets

    source_nets: List[IPv4Network] = nets
    nets_number: int = len(nets)

    while True:
        source_nets, _ = reduce(large_absorb_small, source_nets[1:], ([source_nets[0]], source_nets[0]))
        if len(source_nets) == 1:
            break

        merge_list: List[IPv4Network] = []
        source_nets, _ = reduce(small_merge_large, source_nets[1:], (merge_list, source_nets[0]))
        if len(source_nets) == nets_number or len(source_nets) == 1:
            break

        source_nets = sorted(source_nets)
        nets_number = len(source_nets)

    return source_nets


def get_aggregated(nets: List[IPv4Network]) -> List[str]:
    return [str(n) for n in aggregate_networks(nets)]


def print_result(nets: Result[Error, List[str]]) -> Union[None, NoReturn]:
    if nets.is_success():
        for n in nets.value:
            print(n)
    else:
        print(nets.error)
        exit(1)
    return None


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="""IPv4 networks aggregation.
                                    App aggregates IPv4 networks from given string or file with following mechanic:
                                    1. largest prefix absorbs all its subnet prefixes,
                                      e.g. 10.0.0.0/16 absorbs 10.0.0.0/22, 10.10.0.0/24 and so on;
                                    2. prefixes of the same length merged to their supernet, which prefix is one less,
                                      e.g. 10.0.0.0/24, 10.0.1.0/24 will be merged to 10.0.0.0/23,
                                      but 10.0.0.0/24, 10.0.2.0/24 will not be merged to 10.0.0.0/22,
                                      because their closest supernets (/23) are different
                                      and all merging operations are handled only if supernet prefix is one less.""")
    parser.add_argument("-s", "--string",
                        help="Quoted string of networks separated by space")
    parser.add_argument("-f", "--filepath",
                        help="Path to file which contains networks separated by new line")
    opts = parser.parse_args()

    result = get_nets_from_input(opts.string, opts.filepath
                                ).bind(get_net_from_str
                                ).fmap(get_aggregated)
    print_result(result)

File: tools/ipv4_deduplication_aggregation.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/c-robinson/iplib"
)

type Conf struct {
	str      string
	filePath string
}

var cfg Conf

func filter(s []string, f func(string) bool) []string {
	filtered := make([]string, 0)
	for _, v := range s {
		if f(v) {
			filtered = append(filtered, v)
		}
	}
	return filtered
}

func deduplicate(str []string) []string {
	var list []string

	if len(str) == 0 {
		return str
	}

	set := make(map[string]bool)
	for _, s := range str {
		set[s] = true
	}

	for k := range set {
		list = append(list, k)
	}

	return list
}

func getNetsFromInput(str string, path string) ([]string, error) {
	var list []string
	var err error = nil

	emptyStr := len(strings.TrimSpace(str)) == 0
	emptyPath := len(strings.TrimSpace(path)) == 0

	if emptyStr && emptyPath {
		err = fmt.Errorf("No input defined. Choose string or file input")
		return list, err
	}

	if !emptyStr && !emptyPath {
		err = fmt.Errorf("Both input options can not be used at the same time")
		return list, err
	}

	if !emptyStr {
		list = strings.Split(str, " ")
	} else {
		file, err := os.Open(path)
		if err != nil {
			err = fmt.Errorf("Can not open %s due to %v", path, err)
			return list, err
		}
		defer file.Close()

		reader := bufio.NewScanner(file)
		for reader.Scan() {
			list = append(list, reader.Text())
		}
	}

	list = filter(list, func(s string) bool {
		if len(strings.TrimSpace(s)) == 0 {
			return false
		}
		return true
	})

	list = deduplicate(list)
	return list, err
}

func getNetsFromString(str []string) ([]iplib.Net, error) {
	var list []iplib.Net
	var err error = nil

	for _, s := range str {
		_, n, e := iplib.ParseCIDR(s)
		if e != nil {
			err = fmt.Errorf("Bad IP network value: <%s>", s)
			return list, err
		}
		list = append(list, n)
	}
	// it is very important to keep network list sorted, then absorbing works properly
	sort.Sort(iplib.ByNet(list))
	return list, err
}

// absorbing small net by more large net
// if possible supernet is not real super net for given net, then
// adding the latter to absorbed and make it a new supernet for further nets checking
// otherwise do nothing, just skip the net due to one is absorbed by supernet
// 192.168.0.0/22, 192.168.0.0/24, 192.168.2.0/24 -> 192.168.0.0/22
// Note. Absorbed must be sorted, otherwise it can not work properly
func largeNetsAbsorbSmall(absorbed []iplib.Net, superNet iplib.Net, net iplib.Net) ([]iplib.Net, iplib.Net) {
	if !superNet.ContainsNet(net) {
		absorbed = append(absorbed, net)
		superNet = net
	}
	return absorbed, superNet
}

func lastIsNotGiven(sources []iplib.Net, nets ...iplib.Net) bool {
	last := sources[len(sources)-1]
	result := true
	for _, i := range nets {
		result = result && iplib.CompareNets(last, i) != 0
	}
	return result
}

// compares 2 nets if both have the same maximum closest supernet (e.i. mask - 1), then
// aggregate both to that supernet (merge to large),
// otherwise adding both
// 192.168.0.0/24, 192.168.1.0/24 -> 192.168.0.0/23
func smallMergedToLarge(aggregated []iplib.Net, net1 iplib.Net, net2 iplib.Net) ([]iplib.Net, iplib.Net) {
	net1SuperNet, _ := net1.Supernet(0)
	net2SuperNet, _ := net2.Supernet(0)
	prefixNet1, _ := net1.Mask.Size()
	prefixNet2, _ := net2.Mask.Size()

	// both have the same maximum closest supernet
	if prefixNet1 == prefixNet2 && iplib.CompareNets(net1SuperNet, net2SuperNet) == 0 {
		// if net1 is already in aggregated it has to be deleted if supernet is found
		if len(aggregated) != 0 && !lastIsNotGiven(aggregated, net1) {
			aggregated = aggregated[:len(aggregated)-1]
		}
		aggregated = append(aggregated, net1SuperNet)
		return aggregated, net2
	}

	if len(aggregated) == 0 {
		aggregated = append(aggregated, net1)
		// if last aggregated is not net1 or its maximum closest supernet, then adding one
	} else if lastIsNotGiven(aggregated, net1, net1SuperNet) {
		aggregated = append(aggregated, net1)
	}
	aggregated = append(aggregated, net2)

	return aggregated, net2
}

func aggregateNetworks(nets []iplib.Net) []iplib.Net {
	if len(nets) < 2 {
		return nets
	}
	sourceNets := nets
	netsNumber := len(nets)

	// absorb small networks by more large
	var absorbed []iplib.Net
	superNet := sourceNets[0]
	absorbed = append(absorbed, superNet)
	for _, net := range sourceNets[1:] {
		absorbed, superNet = largeNetsAbsorbSmall(absorbed, superNet, net)
	}
	// no more network for handling, one absorbed all
	if len(absorbed) == 1 {
		return absorbed
	}
	sourceNets = absorbed

	// merge closest networks
	// 192.168.0.0/24, 192.168.1.0/24 -> 192.168.0.0/23
	// 192.168.2.0/24, 192.168.3.0/24 -> 192.168.2.0/23
	// 192.168.0.0/23, 192.168.2.0/23 -> 192.168.0.0/22
	for {
		var mergeList []iplib.Net
		net1 := sourceNets[0]
		for _, net2 := range sourceNets[1:] {
			mergeList, net1 = smallMergedToLarge(mergeList, net1, net2)
		}
		sourceNets = mergeList
		// nothing changed e.i. no closest networks or all megred to one
		if len(sourceNets) == netsNumber || len(sourceNets) == 1 {
			break
		}
		netsNumber = len(sourceNets)
	}

	return sourceNets
}

func printResult(result []iplib.Net) {
	for _, i := range result {
		ip := i.IPNet.IP
		mask, _ := i.Mask.Size()
		fmt.Printf("%v/%v\n", ip, mask)
	}
}

func init() {
	description := `IPv4 networks aggregation.
App aggregates IPv4 networks from given string or file with following mechanic:
1. largest prefix absorbs all its subnet prefixes,
	e.g. 10.0.0.0/16 absorbs 10.0.0.0/22, 10.10.0.0/24 and so on;
2. prefixes of the same length merged to their supernet, which prefix is one less,
	e.g. 10.0.0.0/24, 10.0.1.0/24 will be merged to 10.0.0.0/23,
	but 10.0.0.0/24, 10.0.2.0/24 will not be merged to 10.0.0.0/22,
	because their closest supernets (/23) are different
	and all merging operations are handled only if supernet prefix is one less.
	`
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "%s", description)
		flag.PrintDefaults()
	}
	//getting args
	flag.StringVar(&cfg.str, "string", "", "Quoted string of networks separated by space")
	flag.StringVar(&cfg.filePath, "filepath", "", "Path to file which contains networks separated by new line")
	flag.Parse()
}

func main() {
	exitCode := 0
	defer func() { os.Exit(exitCode) }()

	stringNets, err := getNetsFromInput(cfg.str, cfg.filePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		exitCode = 1
		return
	}

	nets, err := getNetsFromString(stringNets)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		exitCode = 1
		return
	}

	result := aggregateNetworks(nets)

	printResult(result)
}
